File: article/src/main/java/com/cainsgl/article/util/XssSanitizerUtils.java
package com.cainsgl.article.util;

import org.jsoup.safety.Safelist;

/**
 * Markdown XSS 安全过滤工具
 * 
 * 设计原则：
 * 1. 保持 Markdown 原文不变（不转义 ```、#、* 等语法）
 * 2. 只移除嵌入的危险 HTML 标签和属性
 * 3. 不修改换行和空格
 */
public class XssSanitizerUtils {

    // 定义允许的安全 HTML 标签白名单
    private static final Safelist MARKDOWN_SAFELIST = new Safelist()
            // 基础文本标签
            .addTags("p", "br", "span", "div", "hr")
            // 标题
            .addTags("h1", "h2", "h3", "h4", "h5", "h6")
            // 列表
            .addTags("ul", "ol", "li")
            // 文本格式
            .addTags("b", "i", "u", "strong", "em", "strike", "del", "ins", "sub", "sup", "mark")
            // 代码和引用
            .addTags("code", "pre", "blockquote")
            // 表格
            .addTags("table", "thead", "tbody", "tfoot", "tr", "th", "td", "caption")
            // 链接
            .addTags("a")
            .addAttributes("a", "href", "title", "target")
            .addProtocols("a", "href", "http", "https", "mailto")
            // 图片
            .addTags("img")
            .addAttributes("img", "src", "alt", "title", "width", "height")
            .addProtocols("img", "src", "http", "https", "data")
            // 通用属性
            .addAttributes(":all", "class", "id", "style")
            // 表格属性
            .addAttributes("td", "colspan", "rowspan")
            .addAttributes("th", "colspan", "rowspan")
            // 保留换行
            .preserveRelativeLinks(true);

    /**
     * Markdown 内容安全过滤
     */
    public static String sanitize(String markdown) {
        return markdown;
    }


}

File: article/src/main/java/com/cainsgl/article/util/XssSanitizerUtils.kt
package com.cainsgl.article.util

import org.owasp.html.HtmlPolicyBuilder
import org.owasp.html.PolicyFactory

/**
 * XSS防护工具类
 * 使用OWASP Java HTML Sanitizer进行HTML内容清理
 */
object XssSanitizerUtils {
    
    /**
     * HTML清理策略工厂
     * 允许安全的HTML标签和属性，移除潜在的XSS风险
     */
    private val POLICY_FACTORY: PolicyFactory = HtmlPolicyBuilder()
        // 允许的HTML标签
        .allowElements(
            "a", "b", "br", "blockquote", "caption", "cite", "code", "col",
            "colgroup", "dd", "del", "div", "dl", "dt", "em", "h1", "h2",
            "h3", "h4", "h5", "h6", "hr", "i", "img", "kbd", "li", "ol",
            "p", "pre", "q", "samp", "small", "span", "strike", "strong",
            "sub", "sup", "table", "tbody", "td", "tfoot", "th", "thead",
            "tr", "u", "ul"
        )
        // 允许的链接协议
        .allowUrlProtocols("http", "https", "mailto")
        // 允许的a标签属性
        .allowAttributes("href", "title").onElements("a")
        // 允许的img标签属性
        .allowAttributes("src", "alt", "title", "width", "height").onElements("img")
        // 允许的td/th标签属性
        .allowAttributes("colspan", "rowspan").onElements("td", "th")
        // 允许的col标签属性
        .allowAttributes("span", "width").onElements("col")
        // 允许的table标签属性
        .allowAttributes("summary").onElements("table")
        // 允许的blockquote标签属性
        .allowAttributes("cite").onElements("blockquote")
        // 允许的q标签属性
        .allowAttributes("cite").onElements("q")
        // 允许的div/span标签属性
        .allowAttributes("class").onElements("div", "span")
        // 构建策略
        .toFactory()

    /**
     * 清理HTML内容，移除潜在的XSS攻击代码
     * @param htmlContent 包含HTML的内容
     * @return 清理后的安全HTML内容
     */
    fun sanitize(htmlContent: String): String {
        return POLICY_FACTORY.sanitize(htmlContent)
    }
}